openupgradelib/openupgrade.py

```diff
@@ -2606,19 +2608,88 @@ def delete_record_translations(cr, module, xml_ids):
         ),
     )
     for row in cr.fetchall():
-        query = """
-            DELETE FROM ir_translation
-            WHERE module = %s AND name LIKE %s AND res_id = %s;
-        """
-        logged_query(
-            cr,
-            query,
-            (
-                module,
-                row[0] + ",%",
-                row[1],
-            ),
-        )
+        model = row[0]
+        record_id = row[1]
+        if version_info[0] < 16:
+            if not field_list:
+                query = """
+                    DELETE FROM ir_translation
+                    WHERE module = %s AND name LIKE %s AND res_id = %s;
+                """
+                logged_query(
+                    cr,
+                    query,
+                    (
+                        module,
+                        model + ",%",
+                        record_id,
+                    ),
+                )
+            else:
+                name_comparisons = " OR ".join(
+                    ["name = %s" % (model + "," + field) for field in field_list]
+                )
+                query = """
+                    DELETE FROM ir_translation
+                    WHERE module = %s AND ({name_comparison}) AND res_id = %s;
+                """.format(
+                    name_comparison=name_comparisons
+                )
+                logged_query(
+                    cr,
+                    query,
+                    (
+                        module,
+                        record_id,
+                    ),
+                )
+        else:
+            from psycopg2.extras import Json
+
+            table = model.replace(".", "_")
+            # we use information_schema to assure the columns exist
+            cr.execute(
+                """
+                SELECT isc.column_name
+                FROM information_schema.columns isc
+                JOIN ir_model_fields imf ON (
+                    imf.name = isc.column_name AND imf.model = %s)
+                WHERE isc.table_name = %s AND imf.translate""",
+                (model, table),
+            )
+            list_columns = [x[0] for x in cr.fetchall()]
+            if field_list:
+                list_columns = [x for x in list_columns if x in field_list]
+            if not list_columns:
+                continue
+            columns = ", ".join(list_columns)
+            query = """
+                SELECT {columns}
+                FROM {table}
+                WHERE id = %s""".format(
+                table=table,
+                columns=columns,
+            )
+            cr.execute(query, (record_id,))
+            list_translations = cr.fetchall()[0]
+            new_values = []
+            for column, translations in zip(list_columns, list_translations):
+                field_translations = translations or {}
+                if "en_US" in field_translations:
+                    clean_value = {"en_US": field_translations["en_US"]}
```

Comment:
Can we also avoid to read each time the translation, but only to remove extra values. I think JSON SQL allows this.

Reply:
How?

Reply:
And what about to just assign on the UPDATE the existing value? The same you do something like `UPDATE table SET field = field + 1`, you would be able to do something similar with JSON SQL. Invented example: `UPDATE table SET field_json = {"en_US": field_json["en_US"]}`. 

https://www.postgresql.org/docs/current/functions-json.html

Reply:
This contradicts the previous suggestion. Or I read the translations and filter them to avoid doing stupid updates, or I don't read them and just do a complicated update query that will be executed always.

Reply:
Well, more or less. Maybe the SQL `WHERE` can include also a condition to say "if `en_US` is not the only key in the JSON". But I'm fine-tuning a lot, I know.

Reply:
Then, a separate UPDATE query for each column will be done. I cannot batch them into one if we need to check for each one if  they include the en_US key or not.